### DISCUSSION TITLE
"Listen" to actions on a single spawned actor (@xstate/react)

### DISCUSSION THREAD

**Discussion by n8jadams** [statelyai/xstate#2293] [Q&A]
In the docs the example is pretty clear on how to use the `asEffect` hook in conjunction with the `useMachine` hook. (https://xstate.js.org/docs/packages/xstate-react/#aseffect-action)

But I'm in a scenario where I have a machine that's spawning multiple children actors, and I'd like to listen to actions on the child machine, or actor.

On @xstate/react 0.8.1 I've been doing this to listen for an action ('success-saving-animation') from the child machine, and then clearing.

```typescript
import { useRef, useEffect } from 'react'

function useXStateAction(
	service: any,
	actionType: string,
	callback: () => void
): void {
	const latestCallbackRef = useRef(callback)

	useEffect(() => {
		latestCallbackRef.current = callback
	})

	useEffect(() => {
		return service.subscribe(state => {
			if (!state.changed) {
				return
			}

			if (
				!state.actions.find(action => {
					return action.type === actionType
				})
			) {
				return
			}

			latestCallbackRef.current()
		}).unsubscribe
	}, [service, actionType])
}

// And then the component...
function MyChildComponent({ childActorRef }) {
  // The third item here is gone in newest version
  const [state, _, interpreter] = useService(childActorRef)
  const [successAnimation, setSuccessAnimation] = useState(false)
  const clearSavingAnimationRef = useRef<ReturnType<typeof setTimeout>>()
  
  // listens for the action from the child machine...
  useXStateAction(interpreter, 'success-saving-animation', (): void => {
	  setSuccessAnimation(true)
	  clearSavingAnimationRef.current = setTimeout((): void => {
		  setSuccessAnimation(false)
		  clearTimeout(clearSavingAnimationRef.current)
	  }, 3000)
  })

// render code...

}
```

The `interpreter` return (3rd item in the tuple) of `useService` is gone in the newest version of @xstate/react.

How would I go about writing this code now without changing the underlying state machine? (I like the state machine to just fire this action and forget, rather than be concerned about animation timeouts, as I see that as more of a concern of the renderer of the machine, rather than the machine itself.)

**Comment by davidkpiano** [comment]
> The interpreter return (3rd item in the tuple) of useService is gone in the newest version of @xstate/react.

The `interpreter` you're expecting is exactly the same as the `childActorRef` you pass into `useService(...)`; it's redundant.

Anyway, without code, can you describe the use-case of what you're trying to do?

**Reply by n8jadams** [reply]
So basically I swapped out `interpreter` with the `childActorRef` and it worked perfectly.

```typescript
function MyChildComponent({ childActorRef }) {
  const [state] = useService(childActorRef)
  const [successAnimation, setSuccessAnimation] = useState(false)
  const clearSavingAnimationRef = useRef<ReturnType<typeof setTimeout>>()
  
  useXStateAction(childActorRef, 'success-saving-animation', (): void => {
	  setSuccessAnimation(true)
	  clearSavingAnimationRef.current = setTimeout((): void => {
		  setSuccessAnimation(false)
		  clearTimeout(clearSavingAnimationRef.current)
	  }, 3000)
  })

  // render code...

}
```

This is pretty good, but it's still using that custom hook `useXStateAction`, which I'm fine with I guess. I was just hoping to be able to do something like this:

```typescript
function MyChildComponent({ childActorRef }) {
	const [successAnimation, setSuccessAnimation] = useState(false)
	const clearSavingAnimationRef = useRef<ReturnType<typeof setTimeout>>()

	const [state] = useService(childActorRef, {
		actions: {
		 ['success-saving-animation']: () => {
				setSuccessAnimation(true)
				clearSavingAnimationRef.current = setTimeout((): void => {
					setSuccessAnimation(false)
					clearTimeout(clearSavingAnimationRef.current)
				}, 3000)
			}
		})
	})

	// render code...

}
```

And set/override the `success-saving-animation` action for this individual spawned actor... [but you've said before that this is impossible...](https://github.com/davidkpiano/xstate/issues/547#issuecomment-511628096)

The use case is a dynamic form where the user enters information about multiple blog posts. (Think, each blog post has its own title, url, etc.) The parent machine manages some of the communication and adding/removing blog posts from the list, while the children machines manage each individual blog post inputs, error state, etc. The parent machine handles saving all of the children machine's data at once by sending a request to the server, and then after it finishes, it tells the newly updated child machine to show a success animation, which it does so as a side-effect type action, in this case `success-saving-animation`. This parent/child structure is reflected in the react component structure, and the implementation of the `success-saving-animation` is done at the individual child component level.

I don't know if this is confusing... haha, but ultimately I don't want every child component (which is the same type of machine) to show the success message, just the applicable one, so I have to implement the "listener" to this particular child's `success-saving-animation` action in the individual child component.
